README.md
# carrot: 

##

Library to keep your rabbits running...

![alt tag](https://cloud.githubusercontent.com/assets/3204818/23513284/5d24a108-ff5b-11e6-8f0d-12126f820385.png) 

A Clojure library designed to provide you with the implementation of the following simple RabbitMq delayed retry mechanism:
![alt tag](https://cloud.githubusercontent.com/assets/3204818/23512162/99eec068-ff57-11e6-9176-a883f79a9e22.png)


The idea is the following:

1. Provider sends a message to the message queue
2. Consumer tries to process it
3. While processing some exception is thrown
4. Failed message gets into the waiting-queue for a period of time which can be configured (ttl)
5. After the ttl expired the message is put back to message queue to try to process it again
6. Steps 3-5 repeated until the message successfully processed or until the number of retries are less than the max-retry you running carrot with.
7. When we exceed max retry, we put the message in the corresponding dead-letter queue sorted.

## Releases and Dependency Information

* I publish releases to [Clojars]

* Latest release is [carrot "0.1.6"]

* [All releases]

[Leiningen] dependency information:

   [carrot "0.1.6"]

## Usage

No need to worry if the above diagram seems to be too complicated. The idea is that you give custom names to the queues and exchanges you find on the diagram and Carrot will provide you with the retry mechanism and will create the architecturev for you.

Main steps:

- Require carrot in your code:

```clojure
:require [carrot.core :as carrot]
```
- Define your exchange and queue names in a carrot config map (for details see [architecture](https://cloud.githubusercontent.com/assets/3204818/23512162/99eec068-ff57-11e6-9176-a883f79a9e22.png)):

```clojure
(def carrot-config {:waiting-exchange "waiting-exchange"
                    :dead-letter-exchange "dead-letter-exchange"
                    :waiting-queue "waiting-queue"
                    :message-exchange "message-exchange"})
```

- Declare your carrot system which will declare exchanges and queues with the given configuration:
```clojure
 (carrot/declare-system ch
                           carrot-config
                           3000;;ttl spent in waiting queue (in milliseconds)
                           "topic";;type for the exchanges
                           {:durable true};;config for the exchanges
                           )
 ```

- You subscribe for your message queues by using carrot subscribe function:
```clojure
(carrot/subscribe ch
                      carrot-config
                      qname
                      ;;use carrot to create the message handler for langohr:
                      (carrot/crate-message-handler-function
                       (comp
                        message-handler-01
                        message-handler-02
                        message-handler-03
                        ;;here you can en list more functions and they will be threaded in order via threading macro
                        ;;and will compose a message handler function
                        )
                       qname
                       3
                       carrot-config
                       logger)
                      {:auto-ack false};;do not auto acknoledge carrot will do it for you  
                      )
```
[Full example code](src/carrot/examples/example.clj)

## Exception handling
- You  can throw exceptions for not retrying. If you throw this exception it indicates that carrot should not retry the message because the message can't ever be processed, so no reason for retrial.

```clojure
(carrot/throw-do-not-retry-exception "Something fatal happaned." {:cause "The message can not be processed."})
```
[Example code for this](src/carrot/examples/example_with_no_retry_exception.clj)

- Optionally you  can declare functions you don't accept exceptions from, for retrying. Bacically you enlist functions here where any exception means the message can't ever be processed, so no reason for retrial.

```clojure
(carrot/do-not-retry! [#'my-namespace/message-handler-02
                #'my-namespace/message-handler-04])

```
[Example code for this](src/carrot/examples/example_without_retry.clj)

## TODOS
- tests tests and tests
- support strategy for exponential backoff
- support function for replaying messages ended up in dead letter queue
- possibly: provide a function to hanldle dead letters and one strategy could be to put them in dead letter queues

## License

Copyright © 2017 Gabor Raski

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-01-18
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-01-18
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/carrot/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/carrot/compare/0.1.0...0.1.1


test/carrot/core_test.clj
(ns carrot.core-test
  (:require [clojure.test :refer :all]
            [carrot.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject carrot "0.1.6"
  :description "A Clojure library designed to providing the implementation of  RabbitMq delayed retry mechanism."
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :deploy-repositories [["clojars" {:url "https://clojars.org/repo"
                                     :sign-releases false
                                     ;;:signing {:gpg-key "0xAB123456"}
                                    }]]
  :dependencies [[org.clojure/clojure "1.8.0"]
                 ;; rabbitmq client
                 [com.novemberain/langohr "3.6.1"]


                 ;; logging
                 [org.clojure/tools.logging "0.3.1"]
                 [org.slf4j/slf4j-api "1.7.21"]
                 [ch.qos.logback/logback-classic "1.1.7" :exclusions [org.slf4j/slf4j-api]]
                 [ch.qos.logback/logback-access "1.1.7"]
                 [ch.qos.logback/logback-core "1.1.7"]


                 ;; errors
                 [dire "0.5.4"]])


src/carrot/core.clj
(ns carrot.core (:require
            [dire.core :refer [with-handler!]]
            [langohr.basic :as lb]
            [langohr.queue :as lq]
            [langohr.exchange :as le]
            [langohr.consumers :as lc]))

(defn- retry-exception?
  "returns true if the exception has to be retried, false otherwise"
  [exception logger-fn]
  (let [reject? (:reject? (ex-data exception))] ;; reject? = don't retry
    (when (and logger-fn reject?)
       (logger-fn "METRIC ns=carrot.core name=rejecting-exception exception=" exception))
    (not reject?)))

(defn- add-exception-data
  "given an exception e, returns a ClojureInfo exceptionthat contains the extra data d.
  previous data of e should be kept."
  [e d]
  ;; TODO figure out if we can keep the stacktrace as well
  (ex-info (or (.getMessage e) "do not retry me")
           (-> (or (ex-data e) {})
               (merge d))))


(defn- throw-with-extra!
  "given a function f attaches an error handler
  that will catch all exceptions and rethrow them with added data {:reject? true}"
  [f extra]
  (with-handler! f
    Exception
    (fn [e & args]
      (throw (add-exception-data e extra)))))


(defn do-not-retry!
  "attach a supervisor that will not retry the functions in fn-coll"
  [fn-coll]
  (doall (map #(throw-with-extra! % {:reject? true}) fn-coll)))

(defn throw-do-not-retry-exception [msg meta]
  (throw (ex-info msg
                  (assoc meta :reject? true ))))


(defn- nack [ch message meta routing-key retry-attempts max-retry waiting-exchange dead-letter-exchange logger-fn]
  (let [retry-attempts (int retry-attempts)
        exchange (if (> max-retry retry-attempts) waiting-exchange dead-letter-exchange)]
    (when logger-fn (logger-fn "LOGME ns=carrot.core name=current-retry-attempt message:" (:message-id meta) retry-attempts))
    (when logger-fn (logger-fn "Sending message " (:message-id meta)  " to the exchange: " exchange))
    (lb/publish ch
                exchange
                routing-key
                message
                {:persistent true
                 :headers {"retry-attempts" (inc retry-attempts)}}))
  (lb/ack ch (:delivery-tag meta)))

(defn- message-handler [message-handler routing-key max-retry waiting-exchange dead-letter-exchange logger-fn ch meta ^bytes payload]
  (try
    (let [carrot-map {:channel ch
                      :meta meta
                      :payload payload}]
      (-> carrot-map
          message-handler)
      (lb/ack ch (:delivery-tag meta))
      (when logger-fn (logger-fn "LOGME ns=carrot.core name=message-processed message-id=" (:message-id meta))))
    (catch Exception e
      (when logger-fn
        (logger-fn "LOGME ns=carrot.core name=message-process-error message-id="(:message-id meta) "error="e " exception="(clojure.stacktrace/print-stack-trace e)))
      (if (retry-exception? e logger-fn)
        (nack ch payload meta routing-key (or (get (:headers meta) "retry-attempts") 0) max-retry waiting-exchange dead-letter-exchange logger-fn)
        (lb/ack ch (:delivery-tag meta))))))


(defn declare-system [channel
                      {:keys [waiting-exchange dead-letter-exchange waiting-queue message-exchange]}
                      message-ttl
                      exchange-type
                      exchange-config]
  (le/declare channel waiting-exchange exchange-type exchange-config)
  (le/declare channel message-exchange exchange-type exchange-config)
  (le/declare channel dead-letter-exchange exchange-type exchange-config)
  (let [waiting-queue-name (:queue (lq/declare channel waiting-queue
                                 {:exlusive false
                                  :auto-delete true
                                  :durable true
                                  :arguments {"x-message-ttl" message-ttl
                                              "x-dead-letter-exchange" message-exchange}}))]
    (lq/bind channel waiting-queue-name waiting-exchange {:routing-key "#"})))

(defn subscribe [channel
                 {:keys [dead-letter-exchange]}
                 queue-name
                 message-handler
                 queue-config]
  (lc/subscribe channel queue-name message-handler queue-config)
  (lq/declare channel (str "dead-" queue-name) {:exclusive false :auto-delete false})
  (lq/bind channel (str "dead-" queue-name) dead-letter-exchange {:routing-key queue-name}))

(defn crate-message-handler-function
  ([handler routing-key max-retry {:keys [waiting-exchange dead-letter-exchange]} logger-fn]
   (partial message-handler handler routing-key max-retry waiting-exchange dead-letter-exchange logger-fn))
  ([handler routing-key max-retry {:keys [waiting-exchange dead-letter-exchange]}]
   (crate-message-handler-function message-handler handler routing-key max-retry waiting-exchange dead-letter-exchange nil)))

(defmacro compose-payload-handler-function
  [& args]
  (list 'fn ['payload]
          (concat (list '-> 'payload)
                args)))


src/carrot/rabbit_client.clj
(ns carrot.rabbit
  (:require [cheshire.core :as json]
            [cljrc-automod.coerce :as coerce]
            [cljrc-automod.comment :refer [handle-comment handle-facebook]]
            [cljrc-automod.comment.rabbit-constants :as rc]
            [cljrc-automod.facebook]
            [cljrc-automod.report-abuse :refer [handle-abuse-report]]
            [cljrc-automod.schema :as xs]
            [cljrc-automod.vote :refer [handle-voteup
                                        handle-votedown
                                        handle-unvoteup
                                        handle-unvotedown]]
            [clojure.tools.logging :as log]
            [com.stuartsierra.component :refer [Lifecycle]]
            [dire.core :refer [with-handler!]]
            [langohr.basic :as lb]
            [langohr.channel :as lch]
            [langohr.consumers :as lc]
            [langohr.core :as rmq]
            [langohr.exchange :as le]
            [langohr.queue :as lq]
            [perseverance.core :as p]
            [schema.core :as s]))

(defn- retry-exception?
  "returns true if the exception has to be retried, false otherwise"
  [exception]
  (let [reject? (:reject? (ex-data exception))] ;; reject? = don't retry
    (when reject?
      (log/debug "METRIC ns=rc.automod name=rejecting-exception exception=" exception))
    (not reject?)))

(defn add-exception-data
  "given an exception e, returns a ClojureInfo exceptionthat contains the extra data d.
  previous data of e should be kept."
  [e d]
  ;; TODO figure out if we can keep the stacktrace as well
  (ex-info (or (.getMessage e) "do not retry me")
           (-> (or (ex-data e) {})
               (merge d))))

(defn throw-with-extra!
  "given a function f attaches an error handler
  that will catch all exceptions and rethrow them with added data {:reject? true}"
  [f extra]
  (with-handler! f
    Exception
    (fn [e & args]
      (throw (add-exception-data e extra)))))

(defn do-not-retry!
  "attach a supervisor that will not retry the functions in fn-coll"
  [fn-coll]
  (doall (map #(throw-with-extra! % {:reject? true}) fn-coll)))

(do-not-retry! [#'cljrc-automod.comment/add-user-to-comment
                #'cljrc-automod.facebook/call-facebook-feed])

;; TODO: Need to consider how to handle errors. At present the handler simply
;; acks after a single retry. We may wish to move to some queue to retry at a
;; later a date. Also, it would be good log additional info for splunk searches,
(defn- handle-message
  ""
  [handler coerce-fn ch meta ^bytes payload]
  (try
    (let [string (String. payload "UTF-8")
          _ (log/info "METRIC ns=rc.automod name=message-received msg=" string)]
      (-> string
          (json/parse-string true)
          (coerce/clojurize-keys)
          (coerce-fn)
          handler)
      (lb/ack ch (:delivery-tag meta))
      (log/info "METRIC ns=rc.automod name=message-processed msg=" string))
    (catch Exception e
      (if (:redelivery? meta)
        (log/error "METRIC ns=rc.automod name= message-id="(:message-id meta) "error="e "exception="(clojure.stacktrace/print-stack-trace e)) ;; No message acknowledgement.
        (log/error "METRIC ns=rc.automod name=message-first-attempt--error message-id="(:message-id meta) "error="e "exception="(clojure.stacktrace/print-stack-trace e)))
      (if (retry-exception? e)
        (lb/nack ch (:delivery-tag meta) false true)
        (lb/ack ch (:delivery-tag meta))))))

(defn- create-rabbit-connection [rabbit-config]
  (log/info "Rabbit config:" rabbit-config)
  (let [rconn (rmq/connect rabbit-config)
        rch (lch/open rconn)]
    (try
      (le/declare rch rc/message-exchange "topic" {:durable true})
      {:connection rconn :channel rch}
      (catch Exception e
        (log/info "error: " (.getMessage e))
        (when (and rch (.isOpen rch))
          (rmq/close rch))
        (when (and rconn (.isOpen rconn))
          (rmq/close rconn))
        :error))))

(defn create-consumer [ch h routing-key queue-name config]
  (let [qn (:queue (lq/declare ch queue-name
                               {:exlusive false
                                :auto-delete true
                                :durable false
                                :arguments {"x-message-ttl" (or (:message-ttl config) 5000)
                                            "x-dead-letter-exchange" rc/dead-letter-exchange}}))]
    (lq/bind ch qn rc/message-exchange {:routing-key routing-key})
    (lc/subscribe ch qn h {:auto-ack false})))

(defn- assert-mq-settings [{:keys [host consumer-id]}]
  (when (or (empty? host) (empty? consumer-id))
    (throw (Exception. (format "Rabbit-mq config error - ensure :host and :consumer-id are set. Sup(plied values :host - %s, :consumer-id - %s"
                               host
                               consumer-id)))))

(defrecord RabbitMqClient [config]
  Lifecycle

  (start [component]
    ;;(assert-mq-settings config)
    (let [rabbit (create-rabbit-connection config)]
      (doseq [[queue routing-key handler schema]
              [["post.comment" "post.comment" handle-comment xs/RabbitComment]
               ["post.abuse.report" "post.abuse.report" handle-abuse-report xs/RabbitAbuseReport]
               ["post.voteup" "post.voteup" handle-voteup xs/RabbitCommentVote]
               ["post.votedown" "post.votedown" handle-votedown xs/RabbitCommentVote]
               ["post.unvoteup" "post.unvoteup" handle-unvoteup xs/RabbitCommentVote]
               ["post.unvotedown" "post.unvotedown" handle-unvotedown xs/RabbitCommentVote]
               ["post.moderation.facebook" "post.moderation.facebook" handle-facebook xs/RabbitFbComment]]]
        (create-consumer (:channel rabbit) (partial handle-message handler (coerce/parse-rabbit schema)) routing-key queue config))

      (assoc component
        :rabbit-connection (:connection rabbit)
        :rabbit-channel (:channel rabbit))))

  (stop [{:keys [rabbit-connection rabbit-channel :as component]}]
    (when (and rabbit-channel (.isOpen rabbit-channel))
      (rmq/close rabbit-channel))
    (when (and rabbit-connection (.isOpen rabbit-connection))
      (rmq/close rabbit-connection))
    (dissoc component :rabbit-connection :rabbit-channel)))

;; (defn publish-message [msg]
;;   (lb/publish import-broker-ch "" (format "AssetImport.%s" mq-consumer-id)
;;               (json/generate-string {})
;;               {:persistent true}))

(defn make-rabbit-client [config]
  (RabbitMqClient. config))

(comment
  (require 'cljrc-automod.system)

  (lb/publish (:rabbit-channel (cljrc-automod.system/component :rabbit)) "post.comment" "post.comment" (json/generate-string {:comment-id 123})))


src/carrot/examples/example.clj
(ns carrot.examples.example
  (:gen-class)
  (:require [carrot.core :as carrot]
            [langohr.core      :as rmq]
            [langohr.exchange  :as le]
            [langohr.channel   :as lch]
            [langohr.queue     :as lq]
            [langohr.consumers :as lc]
            [langohr.basic     :as lb]
            [clojure.tools.logging :as log]))

(def ^{:const true}
  default-exchange-name "")

;;your custom message handler with exception to test retry mechanism with
(defn message-handler-01
  [{:keys [ch meta payload]}]
  (println (format "[consumer] Received a message: %s"
                   (String. payload "UTF-8"))))

(defn message-handler-02
  [{:keys [ch meta payload], :as carrot-map}]
  (println (format "[consumer] Received a message: %s"
                   (String. payload "UTF-8")))
  (throw (Exception. "my exception for retry message")))

;;ypurlogger function. It is optional
(defn logger [ & all]
  (log/info all))

;;define ypur exchange and queue names in a carrot config map:
(def carrot-config {:waiting-exchange "waiting-exchange"
                    :dead-letter-exchange "dead-letter-exchange"
                    :waiting-queue "waiting-queue"
                    :message-exchange "message-exchange"})



(defn- main
  [& args]
  (let [conn  (rmq/connect)
        ch    (lch/open conn)
        qname "message-queue"]
    (println (format "[main] Connected. Channel id: %d" (.getChannelNumber ch)))
    ;;declare your carrot system
    (carrot/declare-system ch
                           carrot-config
                           3000
                           "topic"
                           {:durable true})

    ;;declare your queue where you want to send your business messages:
    (lq/declare ch qname {:exclusive false :auto-delete false})
    ;;bind your queue to your main message exchange. Use the same name you defined in carrot system config:
    (lq/bind ch qname "message-exchange" {:routing-key qname})
    (carrot/subscribe ch
                      carrot-config
                      qname
                      ;;user carrot to create the message handler for langohr:
                      (carrot/crate-message-handler-function
                       (comp
                        message-handler-01
                        message-handler-02
                        ;;here you can en list more functions and they will be threaded in order via threading macro
                        ;;and will compose a message handler function
                        )
                       qname
                       3
                       carrot-config
                       logger)
                      {:auto-ack false})
    (lb/publish ch default-exchange-name qname "Hello World!" {:content-type "text/plain" :type "greetings.hi"})
    (Thread/sleep 20000)
    (println "[main] Disconnecting...")
    (rmq/close ch)
    (rmq/close conn)))


src/carrot/examples/example_without_retry.clj
(ns carrot.examples.example-without-retry
  (:gen-class)
  (:require [carrot.core :as carrot]
            [langohr.core      :as rmq]
            [langohr.exchange  :as le]
            [langohr.channel   :as lch]
            [langohr.queue     :as lq]
            [langohr.consumers :as lc]
            [langohr.basic     :as lb]
            [clojure.tools.logging :as log]))

(def ^{:const true}
  default-exchange-name "")

;;your custom message handler with exception to test retry mechanism with
(defn message-handler-01
  [{:keys [ch meta payload], :as carrot-map}]
  (println (format "[consumer] Received a message: %s"
                   (String. payload "UTF-8")))
  carrot-map)

(defn message-handler-02
  [{:keys [ch meta payload], :as carrot-map}]
  (println (format "[consumer] Received a message: %s"
                   (String. payload "UTF-8")))
  (throw (Exception. "my exception for retry message")))

(carrot/do-not-retry! [#'message-handler-02])

;;ypurlogger function. It is optional
(defn logger [ & all]
  (log/info all))

;;define ypur exchange and queue names in a carrot config map:
(def carrot-config {:waiting-exchange "waiting-exchange"
                    :dead-letter-exchange "dead-letter-exchange"
                    :waiting-queue "waiting-queue"
                    :message-exchange "message-exchange"})

(defn- main
  [& args]
  (let [conn  (rmq/connect)
        ch    (lch/open conn)
        qname "message-queue"]
    (println (format "[main] Connected. Channel id: %d" (.getChannelNumber ch)))
    ;;declare your carrot system
    (carrot/declare-system ch
                           carrot-config
                           3000
                           "topic"
                           {:durable true})

    ;;declare your queue where you want to send your business messages:
    (lq/declare ch qname {:exclusive false :auto-delete false})
    ;;bind your queue to your main message exchange. Use the same name you defined in carrot system config:
    (lq/bind ch qname "message-exchange" {:routing-key qname})
    (carrot/subscribe ch
                      carrot-config
                      qname
                      ;;user carrot to create the message handler for langohr:
                      (carrot/crate-message-handler-function
                       (comp
                        message-handler-01
                        message-handler-02
                        ;;here you can en list more functions and they will be threaded in order via threading macro
                        ;;and will compose a message handler function
                        )
                       qname
                       3
                       carrot-config
                       logger)
                      {:auto-ack false})
    (lb/publish ch default-exchange-name qname "Hello World!" {:content-type "text/plain" :type "greetings.hi"})
    (Thread/sleep 20000)
    (println "[main] Disconnecting...")
    (rmq/close ch)
    (rmq/close conn)))


src/carrot/examples/example_with_no_retry_exception.clj
(ns carrot.examples.example-with-no-retry-exception
  (:gen-class)
  (:require [carrot.core :as carrot]
            [langohr.core      :as rmq]
            [langohr.exchange  :as le]
            [langohr.channel   :as lch]
            [langohr.queue     :as lq]
            [langohr.consumers :as lc]
            [langohr.basic     :as lb]
            [clojure.tools.logging :as log]))

(def ^{:const true}
  default-exchange-name "")

;;your custom message handler with exception to test retry mechanism with
(defn message-handler-01
  [{:keys [ch meta payload], :as carrot-map}]
  (println (format "[consumer] Received a message part 1: %s"
                   (String. payload "UTF-8")))
  carrot-map)

(defn message-handler-02
  [{:keys [ch meta payload], :as carrot-map}]
  (println (format "[consumer] Received a message part 2: %s"
                   payload "U"))
  (carrot/throw-do-not-retry-exception "Something fatal happaned." {:cause "The message can not be processed."})
  carrot-map)


;;yourlogger function. It is optional
(defn logger [ & all]
  (log/info all))

;;define your exchange and queue names in a carrot config map:
(def carrot-config {:waiting-exchange "waiting-exchange"
                    :dead-letter-exchange "dead-letter-exchange"
                    :waiting-queue "waiting-queue"
                    :message-exchange "message-exchange"})

(defn- main
  [& args]
  (let [conn  (rmq/connect)
        ch    (lch/open conn)
        qname "message-queue"]
    (println (format "[main] Connected. Channel id: %d" (.getChannelNumber ch)))
    ;;declare your carrot system
    (carrot/declare-system ch
                           carrot-config
                           3000
                           "topic"
                           {:durable true})

    ;;declare your queue where you want to send your business messages:
    (lq/declare ch qname {:exclusive false :auto-delete false})
    ;;bind your queue to your main message exchange. Use the same name you defined in carrot system config:
    (lq/bind ch qname "message-exchange" {:routing-key qname})
    (carrot/subscribe ch
                      carrot-config
                      qname
                      ;;user carrot to create the message handler for langohr:
                      (carrot/crate-message-handler-function
                       (comp
                        message-handler-01
                        message-handler-02)
                       qname
                       3
                       carrot-config
                       logger)
                      {:auto-ack false})
    (lb/publish ch default-exchange-name qname "Hello World!" {:content-type "text/plain" :type "greetings.hi"})
    (Thread/sleep 20000)
    (println "[main] Disconnecting...")
    (rmq/close ch)
    (rmq/close conn)))
